finetuner/utils.py

```diff
@@ -0,0 +1,115 @@
+        item that should be semantically close to the first (False). 
+    :return: A generator of :class:`Document`s. Each document represents one line of the
+        input CSV.
+
+    """
+
+    if task == 'any':
+        raise ValueError('MLP model does not support values read in from CSV files.')
+    else:
+        t1, t2 = (
+            task.split('-to-') if len(task.split('-to-')) == 2 else ('text', 'text')
+        )
+        checked = False
+
+    if hasattr(file, 'read'):
+        file_ctx = nullcontext(file)
+    else:
+        file_ctx = open(file, 'r', encoding=encoding)
+
+    with file_ctx as fp:
+        # when set to auto, then sniff
+        try:
+            if isinstance(dialect, str) and dialect == 'auto':
+                dialect = csv.Sniffer().sniff(fp.read(1024))
+                fp.seek(0)
+        except Exception:
+            dialect = 'excel'  #: can not sniff delimiter, use default dialect
+
+        lines = csv.reader(fp, dialect=dialect)
+        artificial_label = 0
+        for col1, col2 in _subsample(lines, size, sampling_rate):
+
+            if not checked:  # determining which column contains images
+                if t1 == 'text' and t2 == 'image':
+                    if _is_uri(col1) and not _is_uri(col2):
+                        t1 = 'image'
+                        t2 = 'text'
+                    elif not _is_uri(col2):
+                        raise ValueError(
+                            (
+                                'uri required in at least one colum ',
+                                'for model with task: ',
+                                task,
+                                '.',
+                            )
+                        )
+                print(t1, t2)
+                checked = True
+            if t1 == 'image':
+                if _is_uri(col1):
+                    d1 = Document(uri=col1, modality='image')
+                else:
+                    raise ValueError(f'Expected uri in column 1, got {col1}')
+            else:
+                d1 = Document(text=col1, modality='text')
+            if is_labeled:
+                label = col2
+                d2 = None
+            elif t2 == 'image':
+                if _is_uri(col2):
+                    d2 = Document(uri=col2, modality='image')
+                else:
+                    raise ValueError(f'Expected uri in column 2, got {col2}')
+            else:
+                d2 = Document(text=col2, modality='text')
+
+            if d2 is None:
+                d1.tags[__DEFAULT_TAG_KEY__] = label
+                yield d1
+            elif (d1.text and d2.text) or (d1.uri and d2.uri):
+                # same modality
+                d1.tags[__DEFAULT_TAG_KEY__] = artificial_label
+                d2.tags[__DEFAULT_TAG_KEY__] = artificial_label
+                artificial_label += 1
+                yield d1
+                yield d2
```

Comment:
can you explain the logic here? the code is very confusing

Reply:
Above this block, the expected modality of each column is determined. Then for each row, the columns are checked to see if their contents are valid for the expected modality (checking if the columns are valid uris on lines 87 and 97), no checks need to be made for text, it is just assigned straight away. If the column is not valid an error is thrown.

when checking the second column, the `is_labeled` option is checked as well, if it is true then the value in the second column is used as a label for the value in the first column, instead of being made into a document itself.

Finally, the two resulting documents are checked to determine what should be returned:
If d2 is none, then d1 is assigned the label and d1 is yielded. If both documents are of the same modality, then both are assigned the same label and are yielded one after the other. Finally, if the document have different modalities, we can infer that this is for a CLIP model, and so they are stored as chunks within the same document and that multimodal document is yielded